### ISSUE TITLE
Add a dataframe viewer/editor to napari (or superqt?)

### ISSUE THREAD

**Issue by goanpeca** [napari/napari#3512]
## 🚀 Feature

Add a dataframe viewer/editor to napari (or superqt?)

![image](https://user-images.githubusercontent.com/3627835/138525932-98666e4a-343d-49ba-921d-902836ae1664.png)

![image](https://user-images.githubusercontent.com/3627835/138525943-32645956-2eb0-43e8-ad3f-cf5d514b15c8.png)

## Motivation

This relates to the work of having a way to edit layer properties or more information eventually. Spyder has a pretty good dataframe editor and we could add this to napari or superqt.

## Pitch

We can not exactly vendor it since it is coupled to Spyder code, so I would suggest superqt and have it be generic enough that napari and other users can use this and other viewers.

## Alternatives

🤷🏼 

## Additional context

References to related code and docs:
- https://docs.spyder-ide.org/current/panes/variableexplorer.html#dataframes
- https://github.com/spyder-ide/spyder/blob/master/spyder/plugins/variableexplorer/widgets/dataframeeditor.py

---

I think @ppwadhwa  start copying (with ll the needed attributions and licensing) over superqt.

Thoughts @napari/core-devs ?

**Comment by alisterburt** [comment]
agreed this would be super nice, definitely a nice quality of life improvement and generic enough for superqt! It could also eventually become an option the magicgui table widget

**Comment by tlambert03** [comment]
Would be useful to have a feature list that you're aiming for here.  Magicgui Table was intended to be this. Accepts dataframes and outputs dataframes ... and has events to connect to for editing

Certainly not completely fully featured but a wishlist would be nice

**Comment by GenevieveBuckley** [comment]
I'm supportive

**Comment by kevinyamauchi** [comment]
I am also +1! The current Magicgui Table works well for most of my use cases. To get things started, I've made a quick list of features below.

**high priority"**
- Inputs: pandas DataFrame and a dict of numpy arrays
- Outputs:  pandas DataFrame and a dict of numpy arrays
- Events: value changed, cell selected, row selected, column selected, others?
- Style: I don't have super concrete thoughts here, but I think it should be easy for the user and devs to change the styling (e.g., fonts and font sizes) since the size of the tables and table values are going to vary widely. For numeric values, I think it would be useful to specify the precision to avoid super wide float columns (perhaps similar to how it is done in formatted strings).

**"nice to have" (in rough order of my priority)**
1. allow the columns to have types (https://github.com/napari/magicgui/issues/216). bonus points for being able to attach a validator function.
2. ability to "split" (e.g., keep the header row in view while the rest of the table scrolls)
2. allow colormaps to be attached to columns to color cells by value
3. lazy loading
4. a "search bar" the user can use to query the table

**Comment by goanpeca** [comment]
Sure @tlambert03, I was not aware of that, so we could expand on that. Some of the nice things that the this dataframe editor has is:

- Lazy loading of data
- Multiindex support (like the image example)

For the rest I thing we could start with that one for sure and add things to it.

---

However even if it lives in magivgui does not mean we have to use it via magic gui correct? we can just import the widget and enhance it/place it as needed in napari?

Also does this fit into magic GUI itself or would it be more of a superqt thing?

>I am also +1! The current Magicgui Table works well for most of my use cases. To get things started, I've made a quick list of features below.

Thanks for chiming it @kevinyamauchi . That list is very helpful. The search bar, or anything that requires an extra widget feels like more of a custom widget to me (built/composed on the napari side?), as in this should be the table itself, unless we provide a hover on finder with assignable shortcuts etc.

**Comment by tlambert03** [comment]
> However even if it lives in magicgui does not mean we have to use it via magicgui correct? we can just import the widget and enhance it/place it as needed in napari?

I'm afraid not:  and this is definitely an important consideration.  A lot of the logic on the table lives outside of the QTable itself... and is backend agnostic, (it also does not depend on pandas):

https://github.com/napari/magicgui/blob/main/magicgui/widgets/_table.py

The Qt part merely satisfies the table protocol required by magicgui: 

https://github.com/napari/magicgui/blob/fee27c6b17b42f6dd422f66328c91e422e19cce6/magicgui/backends/_qtpy/widgets.py#L1061

So, if we strictly want something that goes straight to Qt, then i guess we'll have to start over.
This particular feature (the table) is rather complicated, and probably doesn't fit into the whole backend-agnostic design of magicgui as well as the others do.  So... unfortunate though it may be, I would definitely understand/support a move to start over in pure Qt.

**Comment by goanpeca** [comment]
>So, if we strictly want something that goes straight to Qt, then i guess we'll have to start over.
This particular feature (the table) is rather complicated, and probably doesn't fit into the whole backend-agnostic design of magicgui as well as the others do. So... unfortunate though it may be, I would definitely understand/support a move to start over in pure Qt.

Understood @tlambert03, will take a look at the code and reach out to you with questions :) 

This widget in particular indeed is/can be quite complex indeed.

**Comment by kevinyamauchi** [comment]
> The search bar, or anything that requires an extra widget feels like more of a custom widget to me (built/composed on the napari side?), as in this should be the table itself, unless we provide a hover on finder with assignable shortcuts etc.

Agreed!
